[wasm][debugger] Implement support to Set Next Instruction Pointer.

thaystg opened the pull request:
I'm using "DotnetDebugger.setNextIP" extension of CDP to implement the behavior similar to what we have when "drag the arrow" in a console app.

This functionality is really useful mostly if used together with ApplyChanges. Then you can change the code and then move the execution pointer to a place in the code that is previous to what you change and then re-execute the code changed.

Fixes https://github.com/dotnet/runtime/issues/64475

thaystg committed:
Implement Set Next Instruction Pointer.

thaystg committed:
Merge remote-tracking branch 'origin/main' into thays_implement_set_next_ip

radical commented on src/mono/wasm/debugger/BrowserDebugProxy/MonoProxy.cs:
```diff
@@ -1515,6 +1529,28 @@ private void OnCompileDotnetScript(MessageId msg_id, CancellationToken token)
+
+            var res = new List<SourceLocation>();
+
+            var sourceLocation = new SourceLocation(id, lineNumber, -1);
```
Maybe get the column number from the location too, and if it's missing then default to -1?

radical commented on src/mono/wasm/debugger/BrowserDebugProxy/DebugStore.cs:
```diff
@@ -1189,16 +1193,21 @@ public List<SourceLocation> FindPossibleBreakpoints(SourceLocation start, Source
+
+        public void AddPossibleBreakpointsInMethodToList(SourceLocation start, SourceLocation end, List<SourceLocation> res, MethodInfo method)
+        {
+            if (!method.DebugInformation.SequencePointsBlob.IsNil)
```
This `method.DebugInformation.SequencePointsBlob.IsNil` is being used in multiple places. Maybe it should be turned into a bool property? `method.HasSequencePoints`, or something?

radical commented on src/mono/wasm/debugger/BrowserDebugProxy/MonoProxy.cs:
```diff
@@ -1515,6 +1529,28 @@ private void OnCompileDotnetScript(MessageId msg_id, CancellationToken token)
+            DebugStore store = await RuntimeReady(sessionId, token);
+            ExecutionContext context = GetContext(sessionId);
+            Frame scope = context.CallStack.First<Frame>();
+            if (!SourceId.TryParse(script_id, out SourceId id))
+                return false;
```
This can be done at the start of the method.

radical commented on src/mono/wasm/debugger/BrowserDebugProxy/MonoProxy.cs:
```diff
@@ -1515,6 +1529,28 @@ private void OnCompileDotnetScript(MessageId msg_id, CancellationToken token)
+
+            var sourceLocation = new SourceLocation(id, lineNumber, -1);
+
+            store.AddPossibleBreakpointsInMethodToList(sourceLocation, sourceLocation, res, scope.Method.Info);
```
Should this use the script id to identify the method, instead of using `scope.Method`?

radical replied:
And if it failed to add anything, then should this method fail?

thaystg replied:
I don't think so, we can just move the cursor from the current method being executed. Also scriptId just tells us the file that is being executed and not the method.

radical commented on src/mono/wasm/debugger/BrowserDebugProxy/MonoProxy.cs:
```diff
@@ -1515,6 +1529,28 @@ private void OnCompileDotnetScript(MessageId msg_id, CancellationToken token)
+
+            var ret = await context.SdbAgent.SetNextIP(scope.Method, context.ThreadId, ilOffset, token);
+            var breakpointId = await context.SdbAgent.SetBreakpoint(scope.Method.DebugId, ilOffset.Offset, token);
+            context.BreakpointToDisable = breakpointId;
```
If `context.BreakpointToDisable` is already set, then should it be removed first?

thaystg replied:
I think this cannot happen because we can't continueToLocation twice without waiting for the debugger to be paused in the place that you choose. I mean, it will always call OnContinueToLocation and then the code where we remove the breakpoint before calling OnContinueToLocation again.

radical committed:
Merge remote-tracking branch 'origin/main' into thays_implement_set_next_ip

thaystg commented:
> IIUC, `continue to here` means that the execution resumes, and stops at that line. but here we are changing the IP, thus not executing the code in between.
> 
> What is the "Set Next Instruction Pointer" mentioned in the title supposed to do? If it is meant to change the IP, then like you suggested in the PR description, we should use a different command for it.
> 
> And we can support `continue to here` too, with `Debugger.continueToLocation`.

Okay, I was asking opinions about it. I will create a new message then.

thaystg committed:
Addressing @radical comments.

thaystg committed:
merge

radical commented on src/mono/wasm/debugger/BrowserDebugProxy/MonoProxy.cs:
```diff
@@ -1515,6 +1529,28 @@ private void OnCompileDotnetScript(MessageId msg_id, CancellationToken token)
             SendResponse(msg_id, Result.OkFromObject(new { }), token);
         }
 
+        private async Task<bool> OnSetNextIP(MessageId sessionId, string script_id, int lineNumber, int columnNumber, CancellationToken token)
```
`script_id` -> `scriptId`, to be consistent

radical commented on src/mono/wasm/debugger/DebuggerTestSuite/BreakpointTests.cs:
```diff
@@ -1139,5 +1139,85 @@ public async Task CreateGoodBreakpointAndHitGoToWasmPageWithoutAssetsComeBackAnd
+                "window.setTimeout(function() { invoke_add(); }, 1);",
+                "dotnet://debugger-test.dll/debugger-test.cs", 9, 8,
+                "IntAdd",
+                wait_for_event_fn: (pause_location) =>
```
```suggestion
                wait_for_event_fn: (pause_location) => async (pause_location) =>
```

radical commented on src/mono/wasm/debugger/DebuggerTestSuite/BreakpointTests.cs:
```diff
@@ -1139,5 +1139,85 @@ public async Task CreateGoodBreakpointAndHitGoToWasmPageWithoutAssetsComeBackAnd
+                    CheckNumber(locals, "d", 0);
+                    CheckNumber(locals, "e", 0);
+                    await CheckBool(locals, "f", false);
+                    await Task.CompletedTask;
```
not needed, since we have a `await CheckBool(..`

radical commented:
Add tests for doing this in an async method, and in a lambda too.

thaystg committed:
Checking if the line number is inside the same function.
Adding test for it.
Adding test for moving the IP to a line that has already a breakpoint.

radical commented on src/mono/wasm/debugger/DebuggerTestSuite/BreakpointTests.cs:
```diff
@@ -1139,5 +1139,132 @@ public async Task CreateGoodBreakpointAndHitGoToWasmPageWithoutAssetsComeBackAnd
+                "window.setTimeout(function() { invoke_add(); }, 1);",
+                "dotnet://debugger-test.dll/debugger-test.cs", 9, 8,
+                "IntAdd",
+                wait_for_event_fn: (pause_location) =>
```
Same `async` change needed here too

thaystg replied:
fixed.

thaystg committed:
Creating tests in async function and lambda as suggested by @radical

thaystg commented:
> Add tests for doing this in an async method, and in a lambda too.

Added.

radical commented on src/mono/wasm/debugger/DebuggerTestSuite/SetNextIp.cs:
```diff
@@ -0,0 +1,211 @@
+using Xunit;
+using System.Threading;
+
+namespace DebuggerTests
```
For new files we can use file scoped namespace: `namespace DebuggerTests;`

radical commented on src/mono/wasm/debugger/DebuggerTestSuite/SetNextIp.cs:
```diff
@@ -0,0 +1,211 @@
+namespace DebuggerTests
+{
+
+    public class SetNextIpTests : DebuggerTestBase
```
Add the new class to `eng/testing/scenarios/WasmDebuggerTestsJobsList.txt`.

thaystg committed:
Addressing @radical comments.

thaystg committed:
Addressing @radical comments offline

thaystg committed:
Adding comment to explain dual invocations.

thaystg committed:
Removing unnecessary code.

radical commented on src/mono/wasm/debugger/BrowserDebugProxy/DebugStore.cs:
```diff
@@ -1009,6 +1010,10 @@ public object ToScriptSource(int executionContextId, object executionContextAuxD
                 executionContextId,
                 executionContextAuxData,
                 //hash:  should be the v8 hash algo, managed implementation is pending
+                isLiveEdit = false,
+                isModule = false,
+                scriptLanguage = "JavaScript",
+                hasSourceURL = false,
```
Where are these from?

thaystg replied:
This was needed to implement continueWith, I'll remove this from this PR.

radical commented on src/mono/wasm/debugger/BrowserDebugProxy/MonoProxy.cs:
```diff
@@ -1051,6 +1062,11 @@ private async Task<bool> OnReceiveDebuggerAgentEvent(SessionId sessionId, JObjec
                         Breakpoint bp = context.BreakpointRequests.Values.SelectMany(v => v.Locations).FirstOrDefault(b => b.RemoteId == request_id);
+                        if (request_id == context.BreakpointToDisable)
+                        {
+                                context.BreakpointToDisable = -1;
+                                await context.SdbAgent.RemoveBreakpoint(request_id, token);
```
reduce indent

thaystg committed:
Addressing @radical comments.

radical commented on src/mono/wasm/debugger/BrowserDebugProxy/MonoProxy.cs:
```diff
@@ -499,6 +499,17 @@ protected override async Task<bool> AcceptCommand(MessageId id, string method, J
                 // Protocol extensions
+                case "DotnetDebugger.setNextIP":
+                    {
+                        if (!SourceId.TryParse(args["location"]?["scriptId"]?.Value<string>(), out SourceId sourceId))
+                            return false;
+                        var ret = await OnSetNextIP(id, sourceId, args["location"]["lineNumber"].Value<int>(), args["location"]["columnNumber"].Value<int>(), token);
```
Can we use `SourceLocation.Parse(args["location"])` here? It will check for missing values too.

radical commented on src/mono/wasm/debugger/BrowserDebugProxy/MonoSDBHelper.cs:
```diff
@@ -2683,6 +2683,18 @@ public async Task<bool> SetVariableValue(int thread_id, int frame_id, int varId,
+            commandParamsWriter.Write(method.DebugId);
+            commandParamsWriter.Write((long)ilOffset.Offset);
+            using var getDebuggerCmdReader = await SendDebuggerAgentCommand(CmdThread.SetIp, commandParamsWriter, token);
+            if (getDebuggerCmdReader.HasError)
+                return false;
+            return true;
```
```suggestion
            return !getDebuggerCmdReader.HasError;
```

thaystg committed:
Update src/mono/wasm/debugger/BrowserDebugProxy/MonoSDBHelper.cs

Co-authored-by: Ankit Jain <radical@gmail.com>

radical commented on src/mono/wasm/debugger/DebuggerTestSuite/DebuggerTestBase.cs:
```diff
@@ -475,6 +475,27 @@ internal async Task<Result> SetValueOnObject(JToken obj, string property, string
+                    columnNumber = column
+                });
+
+            if (!expected_error)
+                return await SendCommandAndCheck(
+                    JObject.FromObject(new { location = setNextIPArgs }), "DotnetDebugger.setNextIP", script_loc, line, column, function_name,
+                    wait_for_event_fn: wait_for_event_fn,
+                    locals_fn: locals_fn);
```
```suggestion
            if (!expected_error)
            {
                return await SendCommandAndCheck(
                    JObject.FromObject(new { location = setNextIPArgs }), "DotnetDebugger.setNextIP", script_loc, line, column, function_name,
                    wait_for_event_fn: wait_for_event_fn,
                    locals_fn: locals_fn);
            }
```

radical commented on src/mono/wasm/debugger/DebuggerTestSuite/SetNextIpTests.cs:
```diff
@@ -0,0 +1,144 @@
+
+namespace DebuggerTests;
+
+
```
```suggestion
```

radical commented on src/mono/wasm/debugger/DebuggerTestSuite/SetNextIpTests.cs:
```diff
@@ -0,0 +1,144 @@
+    }
+
+    [Fact]
+        public async Task SetNextIP_Lambda()
+        {
```
```suggestion
    public async Task SetNextIP_Lambda()
    {
```

thaystg committed:
Addressing @radical comments.

thaystg committed:
Apply suggestions from code review

Co-authored-by: Ankit Jain <radical@gmail.com>

radical commented on src/mono/wasm/debugger/BrowserDebugProxy/DevToolsHelper.cs:
```diff
@@ -333,6 +333,8 @@ public ExecutionContext(MonoSDBHelper sdbAgent, int id, object auxData)
 
         private Dictionary<int, PerScopeCache> perScopeCaches { get; } = new Dictionary<int, PerScopeCache>();
 
+        internal int BreakpointToDisable { get; set; }
```
nit: rename to `TempBreakpointForSetNextIP`

thaystg committed:
Removing unnecessary code.Addressing @radical comments.

thaystg committed:
Addressing @radical comments.

radical reviewed (approved):
Thank you for the quick responses, and your patience ;) LGTM 👍

thaystg committed:
Added test as suggested offline by @radical

thaystg committed:
Creating tests as suggested by @radical.

thaystg committed:
Adding test as suggested by @radical.
Fixing behavior of this new test: moving IP to a nested function.

thaystg committed:
Changing order in test.

thaystg committed:
Creating new test suggested by @radical.

radical commented on src/mono/wasm/debugger/BrowserDebugProxy/MonoProxy.cs:
```diff
@@ -1515,6 +1532,50 @@ private void OnCompileDotnetScript(MessageId msg_id, CancellationToken token)
+            var bp = res.First();
+
+            //search if it's a nested function and it's return false because we cannot move to another function
+            if (bp.Line != location.Line || (bp.Line == location.Line && bp.Column != location.Column))
```
```suggestion
            if (bp.Line != location.Line || bp.Column != location.Column))
```

thaystg committed:
Update src/mono/wasm/debugger/BrowserDebugProxy/MonoProxy.cs

Co-authored-by: Ankit Jain <radical@gmail.com>

radical commented on src/mono/wasm/debugger/BrowserDebugProxy/MonoProxy.cs:
```diff
@@ -1515,6 +1532,50 @@ private void OnCompileDotnetScript(MessageId msg_id, CancellationToken token)
+                        continue;
+                    store.AddPossibleBreakpointsInMethodToList(sourceLocation, sourceLocation, res, method);
+                    if (res.Count > 0 &&
+                        (method.StartLocation.Line > scope.Method.Info.StartLocation.Line ||
+                            (method.StartLocation.Line == scope.Method.Info.StartLocation.Line && method.StartLocation.Column > scope.Method.Info.StartLocation.Column)) &&
+                        (method.EndLocation.Line < scope.Method.Info.EndLocation.Line ||
+                            (method.EndLocation.Line == scope.Method.Info.EndLocation.Line && method.EndLocation.Column < scope.Method.Info.EndLocation.Column)))
```
If this is `false` then we don't need to call `AddPossibleBreakpoints*`

thaystg committed:
Addressing @radical comments.

thaystg committed:
Addressing @radical comments.
